commit 5f957ebf94aa5a02815c5bf1e3f59060363efd77
Author: Albert Astals Cid <aacid@kde.org>
Date:   Mon Feb 2 23:59:21 2015 +0100

    Make quoteArgs quote urls with QChar::isSpace()
    
    As David says in the review this is more a workaround than the real fix,
    the real fix would involve modifying splitArgs so that it does not split where it is not needed.
    
    But this is an easier, less impact fix since i don't really know if all shells only split on *real* space or not.
    
    Acked by mpyne
    
    Backported from kcoreaddon b746dcf93239e93eafeedfaeff53e1ac387f5ca1
    
    BuGS: 343380
    FIXED-IN: 4.14.6

diff --git a/kdecore/tests/kshelltest.cpp b/kdecore/tests/kshelltest.cpp
index 1b46f7c..423b61d 100644
--- a/kdecore/tests/kshelltest.cpp
+++ b/kdecore/tests/kshelltest.cpp
@@ -36,6 +36,8 @@ class KShellTest : public QObject
     void quoteArg();
     void joinArgs();
     void splitJoin();
+    void quoteSplit();
+    void quoteSplit_data();
     void abortOnMeta();
 };
 
@@ -155,6 +157,26 @@ KShellTest::splitJoin()
 }
 
 void
+KShellTest::quoteSplit_data()
+{
+    QTest::addColumn<QString>("string");
+
+    QTest::newRow("no space") << QString("hiho");
+    QTest::newRow("regular space") << QString("hi there");
+    QTest::newRow("special space") << QString::fromUtf8("如何定期清潔典型的電風扇　講義.pdf");
+}
+
+void
+KShellTest::quoteSplit()
+{
+    QFETCH(QString, string);
+
+    // Splitting a quote arg should always just return one argument
+    const QStringList args = KShell::splitArgs(KShell::quoteArg(string));
+    QCOMPARE(args.count(), 1);
+}
+
+void
 KShellTest::abortOnMeta()
 {
     KShell::Errors err1 = KShell::NoError, err2 = KShell::NoError;
diff --git a/kdecore/util/kshell_unix.cpp b/kdecore/util/kshell_unix.cpp
index 8614424..f7a4768 100644
--- a/kdecore/util/kshell_unix.cpp
+++ b/kdecore/util/kshell_unix.cpp
@@ -279,7 +279,7 @@ inline static bool isSpecial( QChar cUnicode )
     }; // 0-32 \'"$`<>|;&(){}*?#!~[]
 
     uint c = cUnicode.unicode ();
-    return (c < sizeof(iqm) * 8) && (iqm[c / 8] & (1 << (c & 7)));
+    return ((c < sizeof(iqm) * 8) && (iqm[c / 8] & (1 << (c & 7)))) || cUnicode.isSpace();
 }
 
 QString KShell::quoteArg( const QString &arg )
